File: 1923-sentence-similarity-iii/sentence-similarity-iii.cpp
class Solution {
public:
    bool areSentencesSimilar(string sentence1, string sentence2) {
        stringstream ss1(sentence1), ss2(sentence2);
        string word;
        vector<string> s1Words, s2Words;
        while (ss1 >> word) s1Words.push_back(word);
        while (ss2 >> word) s2Words.push_back(word);

        if (s1Words.size() >= s2Words.size()) {
            swap(s1Words, s2Words);
        }

        int start = 0;
        int s1end = s1Words.size() - 1;
        int s2end = s2Words.size() - 1;

        // finding max number of words matching from start
        while (start < s1Words.size() && s1Words[start] == s2Words[start]) {
            ++start;
        }

        // finding max number of words matching in end
        while (s1end >= 0 && s1Words[s1end] == s2Words[s2end]) {
            --s1end;
            --s2end;
        }

        return s1end < start;
    }
};

File: 1923-sentence-similarity-iii/sentence-similarity-iii.py
class Solution:
    def areSentencesSimilar(self, sentence1: str, sentence2: str) -> bool:
        parts1 = sentence1.split()
        parts2 = sentence2.split()
        
        # parts1 will be the one with lesser words
        if len(parts1) >= len(parts2):
            parts1, parts2 = parts2, parts1

        if len(parts1) == 1:
            return parts1[0] == parts2[0] or parts1[0] == parts2[-1]
        
        # 2 pointers for both sentences
        l1, r1 = 0, len(parts1)-1
        l2, r2 = 0, len(parts2)-1

        while l1 <= r1:
            if parts1[l1] != parts2[l2] and parts1[r1] != parts2[r2]:
                return False
            elif parts1[l1] == parts2[l2] and parts1[r1] == parts2[r2]:
                l1 += 1
                l2 += 1
                r1 -= 1
                r2 -= 1
                continue
            elif parts1[l1] == parts2[l2]:
                l1 += 1
                l2 += 1
            else:
                r1 -= 1
                r2 -= 1
        
        return True